### PR TITLE
proxy/trojan: fix writing UDP packet

### PR DESCRIPTION
**Pull request by dyhkwong**
#2440 #1795

Every UDP packet should have some headers.

### CODE REVIEW

**Comment by Vigilans** [comment]
Ref https://github.com/v2fly/v2ray-core/pull/2336#issuecomment-1435247532, maybe related? 

> 3. In fact, the author made attempt to deal with the issue mentioned above in trojan outbound: 
> https://github.com/v2fly/v2ray-core/blob/83ea1ba3ba94c944d99cb063bc6c5f42d6cf2fff/proxy/trojan/protocol.go#L102-L114
> It **1)** Creates a buffer of length 8192 (maxLength = 8192); **2)** Read the MultiBuffer 8192 bytes by 8192 bytes, then encode each (at most 8192 bytes) buffer into trojan UDP payloads. 
> 
>    This in a sense fixes the problem that trojan inbound writes the UDP payload 2048 bytes by 2048 bytes. Since a UDP payload will not exceed 8192 bytes (max 4096 bytes as you mentioned), the split UDP packet in trojan inbound will be reunited in trojan outbound. 
> 
>    However, even if scoped in trojan inbound -> trojan outbound, this solution is still problematic. How to ensure that the passed `mb buf.MultiBuffer` only contains payload of 1 UDP packet? In fact this is not guaranteed, so the UDP packet in outbound may contains payload of more than 1 UDP packets from inbound, causing misbehavior.

**Comment by Vigilans** [comment]
> > Ref [#2336 (comment)](https://github.com/v2fly/v2ray-core/pull/2336#issuecomment-1435247532), maybe related?
> 
> Indeed. So, this PR also can't handle UDP payload larger than 2048 bytes. Can this be fixed by modifying Trojan's `PacketWriter` and `PacketReader` to make use of `buf.NewWithSize()`, like `packetstream` in #2337?

I think it will fix. I remembered that I was referencing Trojan's PacketWriter and `PacketReader` to implement `packetstream` ones, though I did not have the time and usecase to fix and test for trojan.

File: proxy/trojan/protocol.go [hunk 1]
```diff
@@ -136,15 +135,22 @@ func (w *PacketWriter) WriteTo(payload []byte, addr gonet.Addr) (int, error) {
...
+	}
+
+	buffer := buf.NewWithSize(addrPortBuf.Len() + 2 + 2 + int32(length))
+	defer buffer.Release()
+
+	if _, err := buffer.Write(addrPortBuf.Bytes()); err != nil {
 		return 0, err
 	}
+	addrPortBuf.Release()
 	if _, err := buffer.Write(lengthBuf[:]); err != nil {
```
Comment (Vigilans) [inline proxy/trojan/protocol.go:154]:
```
	if _, err := buffer.Write(addrPortBuf.Bytes()); err != nil {
		return 0, err
	}
	addrPortBuf.Release()
```
If above branch returns, `addrPortBuf` is not released, may cause memory leak?

File: proxy/trojan/protocol_test.go [hunk 2]
```diff
@@ -57,11 +57,12 @@ func TestUDPRequest(t *testing.T) {
...
 	}
+
 	payload := []byte("test string")
-	data := buf.New()
+	data := buf.NewWithSize(int32(len(payload)))
 	common.Must2(data.Write(payload))
 
-	buffer := buf.New()
+	buffer := buf.NewWithSize(int32(len(payload)))
 	defer buffer.Release()
```
Comment (Vigilans) [inline proxy/trojan/protocol_test.go:66]: May add test cases of https://github.com/v2fly/v2ray-core/issues/2440 and udp packets larger than 2048 bytes, so to ensure these scenarios are taken into consideration in future development?

**Review by xiaokangwang — APPROVED** [review]
I have summarily approved this merge request. Thanks for your contribution.

It is worth noting that UDP packet requires segmentation is not widely supported and should be considered edge case.

**Pull request merged**